ArduCopter/mode_auto.cpp

```diff
@@ -131,8 +130,20 @@ void ModeAuto::run()
         break;
 
     case SubMode::CIRCLE:
-        circle_run();
-        break;
+        // check in which state are we in
+        switch (loiter_turns.get_state()) {
+        case _LoiterTurns::State::CIRCLING: {
+            loiter_turns.circle_run();
+            break;
+        }
+        case _LoiterTurns::State::MOVING_TO_EDGE:
+            // use waypoint navigation controller in this state
+            wp_run();
+            break;
+        case _LoiterTurns::State::READY:
+            break;
+        }
+        break; // SubMode::CIRCLE
```

Comment:
Why should auto need to know anything about the internal state?

Can it just call `update()`?

Reply:
Done!